firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,131 @@
+
+/* write parse roundtrip
+Scenario: CRCElement correctly preserves data in roundtrip writing and parsing
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
+      # write  
+      THEN("the crc accessor method returns 0 before the first write method is called")
+      THEN("The first write method call returns ok status")
+      THEN("After the first write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+      # write  
+      THEN("The second write method call returns ok status")
+      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+
+/* CRCElementReceiver
+Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them
+
+  GIVEN("A CRC element receiver of capacity 254 bytes")
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload buffer updated in the CRCElement is independent of the input body")
+
+    WHEN("A body without a complete 4-byte header and empty payload is given to the crc element receiver")
+      THEN("the transform status is equal to invalid parse")
+      THEN("the value returned by the crc accessor method is equal to 0x00000000r")
+      THEN("The body's payload section is empty")
```

Comment:
This doesn't seem correct to me - the body is exactly the contents of input_buffer, which is const. So the payload section of the body is preserved at what it was. Separately, the value returned by the payload accessor method of output_crcelement will be empty if (and only if) output_crcelement had an empty payload before the CRCElementReceiver::transform was called.

Reply:
Yes i'm not sure how this got here, I'm actually not checking what payload accessor returns for this case as it's empty also the `input_buffer` is empty to begin with.

Secondly in all CRCElementReceiver test cases I'm indeed using an empty payload buffer for the output_crcelement. 
So should I get rid of this THEN completely or should I do something like
THEN("The payload buffer returned by the payload accessor method is empty") ?

Reply:
Added.